Adding links to courses in the Django admin (#684)

ssciolla opened the pull request:
This PR makes changes to the Django back-end and `start.sh` in order to provide convenient links in the Django admin to course landing pages in the MyLA front-end. Besides making changes to `models.py` and `admin.py`, I created a new environment variable `LOCALHOST_PORT` and added a new management command `site` so that the domain could be set automatically through `start.sh`. I have tested this locally and on [the MyLA development instance](https://dev-myla.tl.it.umich.edu), where the changes are still deployed (as of the morning of 10/14). More details on the changes are provided below. This PR aims to resolve issue #684.

1) Adding a "View on Site" link to the Course edit page and a field link in the Course list view was relatively straightforward. The former was created automatically once `get_absolute_url` was added as a method to the `Course` model. I then created the list view link by adding a `course_link` method to the `CourseAdmin` model, which could then be included in the `list_display` link. Using [`format_html`](https://docs.djangoproject.com/en/2.2/ref/utils/) seems to be a best practice for preparing small HTML fragments.

2) However, because calling `reverse` pulls in the value for the domain from the `django_site` table in the MySQL database (which was previously `example.com`), I needed a flexible way to set the domain that would work in all the environments the application is run in. The solution I arrived at (there may be others) was to create a new `site` management command that could be invoked from `start.sh`. The new command allows for either the creation of a new site record (using the `--new` flag) or the updating of the default site, which is identified by importing the `SITE_ID` variable from `settings.py`.

3) To isolate the inputs for the new `site` command within `start.sh`, I needed to create a new environment variable `LOCALHOST_PORT` and filter the existing `ALLOWED_HOSTS` array in the `env.json` file. I could not find a place where the port used with `localhost` was specified except in `docker-compose.yml`, and it seemed strange to pull the value from there. Note: for the links to work, reviewers will need to set `LOCALHOST_PORT` if their port differs from the default value of `5001`. I used a [jq](https://stedolan.github.io/jq/manual/v1.5/) query to filter `ALLOWED_HOSTS` down to either a default `localhost` or whatever domain is being used in a deployed instance. Then the `site` command is invoked using either `localhost` and the specified port or simply the provided domain.

Feedback is welcome, and to ensure this works for other institutions, it would be great if someone from UBC could review and test this.

jonespm commented on start.sh:
```diff
@@ -27,20 +27,26 @@ else
     CRONTAB_SCHEDULE=$(jq -r -c ".CRONTAB_SCHEDULE | values" ${ENV_FILE})
     RUN_AT_TIMES=$(jq -r -c ".RUN_AT_TIMES | values" ${ENV_FILE})
+    DOMAIN=$(jq -r -c "${DOMAIN_JQ} | values" ${ENV_FILE})
+    PORT=$(jq -r -c ".LOCALHOST_PORT | values" ${ENV_FILE})
```
I'd probably set a default for this PORT incase it isn't defined in the configuration file. Just put it up above where the other defaults are. It's only used in the localhost case anyway. I guess you could put it in the jquery too.

ssciolla replied:
How does [this commit](https://github.com/tl-its-umich-edu/my-learning-analytics/pull/761/commits/2d63c9fb54edb95c31ee13411acd8fa1992fa954) look? I tried putting the `if fi` block above this, but it was resulting in `PORT` having an empty string value.

jonespm replied:
I think we can forget about this, if we move this from the environment to the docker-compose is should always be set on the localhost.

zqian commented on dashboard/management/commands/site.py:
```diff
@@ -0,0 +1,25 @@
+    def handle(self, *args, **options):
+        domain = options.get('domain')
+        name = options.get('name')
+        new = options.get('new')
```
when will the "new" option be used? Did not see an example in start.sh. May need documentation here.

ssciolla replied:
It isn't being used, but it seemed like a site command (that could be run from the container shell) should be a bit more flexible than just updating the default site. Do you want me to add a comment?

ssciolla replied:
See latest commit, and let me know if that addresses your concern.

zqian replied:
Perfect!

jonespm commented on config/env_sample.json:
```diff
@@ -9,8 +9,9 @@
         "localhost"
     ],
+    "/* Port in use when the application is running on localhost": "*/",
+    "LOCALHOST_PORT": 5001,
```
Inside the container this is running on port 5000 by default. However docker-compose.yml is exposing this as 5001. We *could* switch docker-compose to be 5000 and then not have to have this extra environment variable. However if someone switched it, they'd have to manually update the port somehow. The container doesn't know which port is being exposed on the host. 

We *could* add it to the docker-compose.yml environment instead I think, maybe that's a better idea. Nobody probably needs to configure this in production.

ssciolla commented:
@jonespm, @zqian, okay, I've moved the variable to `docker-compose.yml`. I removed the lines that handle setting that variable from `start.sh`, since the variable will now always be set in the same place and won't need to be configured. Please let me know if you think those deletions are okay.

ssciolla commented:
I just added a new commit trying to address @zqian's feedback that includes comments explaining how the `--new` flag works. I also simplified that argument (using the `--superuser` flag from `createuser` as a model) so that you only need to provide `--new` (as opposed to `--new=True`).
